### PR TITLE
feat: use llamacloud pipeline for private files and generate script

### PR DESCRIPTION
**Pull request by marcusschiesser**
## Summary by CodeRabbit

- **New Features**
  - Enhanced file upload process to include the filename alongside the file data.
  - Improved file processing capabilities, allowing better tracking of file metadata.
  - Introduced new methods for managing files within the LlamaCloud ecosystem, including uploading and downloading files associated with pipelines.
  - Streamlined input handling by shifting from environment variable dependency to direct file processing.

- **Bug Fixes**
  - Strengthened error handling in the file upload component to ensure user feedback is provided even when specific error handlers are not set.

- **Refactor**
  - Streamlined file handling by transitioning from base64 string representation to direct `File` object usage for uploads, improving efficiency and clarity.
  - Conditionally defined API endpoints to improve robustness based on environment configuration.

### CODE REVIEW

File: templates/types/streaming/fastapi/app/api/services/file.py [hunk 1]
```diff
@@ -78,25 +81,25 @@ def store_and_parse_file(file_data, extension) -> List[Document]:
...
         file_data, extension = PrivateFileService.preprocess_base64_file(base64_content)
-        documents = PrivateFileService.store_and_parse_file(file_data, extension)
-
-        # Only process nodes, no store the index
-        pipeline = IngestionPipeline()
-        nodes = pipeline.run(documents=documents)
 
         # Add the nodes to the index and persist it
+        # TODO: get params to select index
         current_index = get_index()
```
Comment (marcusschiesser) [inline templates/types/streaming/fastapi/app/api/services/file.py:89]: @thucpn here we're not passing the index selected in the UI

Comment (marcusschiesser) [reply templates/types/streaming/fastapi/app/api/services/file.py]: @thucpn great catch - the problem is actually that we import LlamaCloudFileService every time - I think the code should not even exist if we don't select LlamaCloud - I'll have a look

File: templates/types/streaming/fastapi/app/api/services/file.py [hunk 2]
```diff
@@ -78,25 +81,30 @@ def store_and_parse_file(file_data, extension) -> List[Document]:
...
+                    custom_metadata={
+                        # Set private=true to mark the document as private user docs (required for filtering)
+                        "private": "true",
+                    },
+                )
+            ]
         else:
-            # Only process nodes, no store the index
+            # First process documents into nodes
+            documents = PrivateFileService.store_and_parse_file(file_data, extension)
```
[the remark this answers is not in the document]
Comment (marcusschiesser) [reply templates/types/streaming/fastapi/app/api/services/file.py:107]: good idea, you can do that in TS already if you want

File: templates/types/streaming/fastapi/app/api/services/llama_cloud.py [hunk 3]
```diff
@@ -41,74 +40,87 @@ def get_all_projects_with_pipelines(cls) -> List[Dict[str, Any]]:
...
-        url = f"{cls.LLAMA_CLOUD_URL}/files/{file_id}/content?project_id={project_id}"
-        return cls._make_request(url)
+    def add_file_to_pipeline(
+        cls,
+        pipeline_id: str,
+        upload_file: Union[typing.IO, Tuple[str, BytesIO]],
+        custom_metadata: Optional[Dict[str, PipelineFileCreateCustomMetadataValue]],
+    ) -> str:
+        client = get_client()
+        file = client.files.upload_file(upload_file=upload_file)
```
[the remark this answers is not in the document]
Comment (marcusschiesser) [reply templates/types/streaming/fastapi/app/api/services/llama_cloud.py]: yes good idea, will do that in python

File: templates/types/streaming/fastapi/app/api/services/llama_cloud.py [hunk 4]
```diff
@@ -41,74 +40,87 @@ def get_all_projects_with_pipelines(cls) -> List[Dict[str, Any]]:
...
-            r.raise_for_status()
-            with open(local_file_path, "wb") as f:
-                for chunk in r.iter_content(chunk_size=8192):
-                    f.write(chunk)
-        logger.info("File downloaded successfully")
+        # Wait 2s for the file to be processed
+        max_attempts = 20
+        attempt = 0
+        while attempt < max_attempts:
+            result = client.pipelines.get_pipeline_file_status(pipeline_id, file.id)
```
[the remark this answers is not in the document]
Comment (marcusschiesser) [reply templates/types/streaming/fastapi/app/api/services/llama_cloud.py]: @thucpn it's a different method - this one here is getting the file status

[the remark this answers is not in the document]
Comment (marcusschiesser) [reply templates/types/streaming/fastapi/app/api/services/llama_cloud.py]: https://docs.cloud.llamaindex.ai/API/get-pipeline-file-status-api-v-1-pipelines-pipeline-id-files-file-id-status-get

File: templates/components/vectordbs/python/llamacloud/generate.py [hunk 5]
```diff
@@ -1,48 +1,46 @@
...
-
+from llama_index.core.readers import SimpleDirectoryReader
+from app.engine.service import LLamaCloudFileService
 
 logging.basicConfig(level=logging.INFO)
 logger = logging.getLogger()
 
 
 def generate_datasource():
-    init_settings()
```
Comment (leehuwuj) [inline templates/components/vectordbs/python/llamacloud/generate.py:17]: Why did you remove `init_settings()`?

Comment (marcusschiesser) [reply templates/components/vectordbs/python/llamacloud/generate.py:17]: not needed as it's using the settings from LlamaCloud

**Pull request merged**